opentelemetry-sdk/src/metrics/internal/sum.rs

```diff
@@ -19,89 +20,122 @@ use super::{aggregate::is_under_cardinality_limit, AtomicTracker, Number};
-            // Try incoming order first
-            if let Some(value_to_update) = values.get(attrs) {
-                value_to_update.add(measurement);
-            } else {
-                // Then try sorted order.
-                let sorted_attrs = AttributeSet::from(attrs).into_vec();
-                if let Some(value_to_update) = values.get(sorted_attrs.as_slice()) {
-                    value_to_update.add(measurement);
-                } else {
-                    // Give up the lock, before acquiring write lock.
-                    drop(values);
-                    if let Ok(mut values) = self.values.write() {
-                        // Recheck both incoming and sorted after acquiring
-                        // write lock, in case another thread has added the
-                        // value.
-                        if let Some(value_to_update) = values.get(attrs) {
-                            value_to_update.add(measurement);
-                        } else if let Some(value_to_update) = values.get(sorted_attrs.as_slice()) {
-                            value_to_update.add(measurement);
-                        } else if is_under_cardinality_limit(self.count.load(Ordering::SeqCst)) {
-                            let new_value = T::new_atomic_tracker();
-                            new_value.add(measurement);
-                            let new_value = Arc::new(new_value);
-
-                            // Insert original order
-                            values.insert(attrs.to_vec(), new_value.clone());
-
-                            // Insert sorted order
-                            values.insert(sorted_attrs, new_value);
-
-                            self.count.fetch_add(1, Ordering::SeqCst);
-                        } else if let Some(overflow_value) =
-                            values.get(STREAM_OVERFLOW_ATTRIBUTES.as_slice())
-                        {
-                            overflow_value.add(measurement);
-                        } else {
-                            let new_value = T::new_atomic_tracker();
-                            new_value.add(measurement);
-                            values.insert(STREAM_OVERFLOW_ATTRIBUTES.clone(), Arc::new(new_value));
-                            global::handle_error(MetricsError::Other("Warning: Maximum data points for metric stream exceeded. Entry added to overflow. Subsequent overflows to same metric until next collect will not be logged.".into()));
-                        }
-                    }
-                }
-            }
+impl<T: Number<T>, O: Operation> ValueMap<T, O> {
+    fn measure(&self, measurement: T, attributes: &[KeyValue]) {
+        if attributes.is_empty() {
+            O::update_tracker(&self.no_attribute_tracker, measurement);
+            self.has_no_attribute_value.store(true, Ordering::Release);
+            return;
+        }
+
+        let Ok(trackers) = self.trackers.read() else {
```

Comment:
thanks for reorganizing this! One of the next steps (in another PR) is to add internal metrics whenever we are dropping a measurement.